### PR TITLE
check mt path before generating the proof

### PR DESCRIPTION
**Pull request by zhenfeizhang**
Added a subroutine to check the validity for merkle path in debug mode; and propagate the **right** error message accordingly.

### CODE REVIEW

File: src/utils/mod.rs [hunk 1]
```diff
@@ -616,6 +617,28 @@ pub(crate) mod txn_helpers {
                 "Fee input and fee change records' public key do not match".to_string(),
             ));
         }
+        // Check that the merkle path is valid w.r.t. ro, path and root.
+        // Only turned on in debugging mode.
+        #[cfg(debug_assertions)]
```
Comment (tri-joe) [inline src/utils/mod.rs:622]: Why only do this check in debug mode? I'd consider it to be an error if there are _any_ inputs to `generate` which yield `WrongQuotientPolyDegree` -- all of that should be checked before running the proof, and should always give better error messages than that.

**Review by chancharles92 — APPROVED** [review]
Thanks, LGTM

**Review by tri-joe — APPROVED** [review]

File: src/utils/mod.rs (hunk 1 above)
Comment (zhenfeizhang) [reply src/utils/mod.rs:622]: I guess @chancharles92 's concern was that the cost of this check will be non-negligible.
Also I believe the idea is that the caller should have checked the validity of `fee_info` before calling `generate` to build the proof.

Agree that error propagation should have been handled better than `WrongQuotientPolyDegree`.

Comment (zhenfeizhang) [reply src/utils/mod.rs]: > Hmm, that's true, if the input is incorrect, it's always more efficient to exit early and output the correct error message. Maybe we can remove the debug mode? @zhenfeizhang

alright. Now it is removed.

**Review by chancharles92 — APPROVED** [review]

**Pull request merged**